elasticdl/python/worker/task_data_service.py

```diff
@@ -27,44 +32,77 @@ def __init__(
             self.data_reader = create_data_reader(data_origin=None)
         self._warm_up_task = None
         self._has_warmed_up = False
+        self._record_failure_tolerance_percentage = (
+            record_failure_tolerance_percentage
+        )
 
     def _reset(self):
         """
         Reset pending tasks and record counts
         """
-        self._record_count = 0
-        self._reported_record_count = 0
-        self._pending_tasks_with_counts = []
+
+        # Counters are keyed by task id.
+        self._cur_record_count = defaultdict(int)
+        self._reported_record_count = defaultdict(int)
+        self._failed_record_count = defaultdict(int)
+        self._pending_tasks = deque()
         self._current_task = None
 
     def get_current_task(self):
         return self._current_task
 
+    def report_task_done(self, task_id, err_msg):
+        self._worker.report_task_result(task_id, err_msg)
+
+    def handle_next_task(self, err_msg=""):
+        with self._lock:
+            task = self._pending_tasks.popleft()
+            self._reported_record_count.pop(task.task_id, None)
+            self._failed_record_count.pop(task.task_id, None)
+            if self._pending_tasks:
+                self._current_task = self._pending_tasks[0]
+            self.report_task_done(task.task_id, err_msg)
+
     def report_record_done(self, count, err_msg=""):
         """
         Report the number of records in the latest processed batch,
         so TaskDataService knows if some pending tasks are finished
         and report_task_result to the master.
-        self._pending_tasks_with_counts[0][0] is the first pending task,
-        self._pending_tasks_with_counts[0][1] is the number of records
-        in this task.
         """
-        self._reported_record_count += count
-        if (
-            len(self._pending_tasks_with_counts)
-            and self._reported_record_count
-            >= self._pending_tasks_with_counts[0][1]
-        ):
-            with self._lock:
-                while (
-                    len(self._pending_tasks_with_counts)
-                    and self._reported_record_count
-                    >= self._pending_tasks_with_counts[0][1]
+        if self._pending_tasks:
+            task = self._pending_tasks[0]
+            self._reported_record_count[task.task_id] += count
```

Comment:
No need to define a dict for `_reported_record_count`. Just define an attribute `self._ _reported_record_count_in_pending_tasks`, and subtract the finished pending task record count in  `handle_next_task`.

Reply:
Fixed